Reflection support for type hints and return types

mbeccati opened the pull request:
A work based off of Sara's https://wiki.php.net/rfc/reflectionparameter.typehint with a twist.

I've changed ReturnTypeAnnotation back to ReturnTypeHint, as this is still what we're calling them.

The new methods are:
* ReflectionFunctionAbstract::hasReturnTypeHint()
* ReflectionFunctionAbstract::getReturnTypeHint()
* ReflectionTypeHint::isInstance() 
* ReflectionTypeHint::isScalar() 

References:
http://markmail.org/message/meyx6t52opv7xfbo

Btw, not trying to bypass the RFC process, but atm I don't have time to propose one.

smalyshev commented:
Please do not propagate the misnomer "type hint". There's nothing of "hint" in type definition. We've made this mistake once, but we should not make it in perpetuity. We should phase out broken terminology instead.

Kubo2 commented:
@smalyshev:

So how would you call it then? The thing _it_ accomplishes isn't actually done by presence of an annotation in code (or do I understand it wrong?).

mbeccati commented:
@smalyshev Sure. However, I thought we settled for "hint" though, as that's the jargon that's been used extensively during the recent Scalar Type Hints civil war ;)

nikic commented:
Looks okay with TypeHint changed back to TypeAnnotation. @morrisonlevi Thoughts?

mbeccati commented:
@nikic isn't TypeAnnotation a bit too close to Annotations, which aren't supported yet, but something very different?

nikic commented:
@mbeccati I don't really think that there's much potential for confusion in that direction. But, maybe just call it `ReflectionType`? Would also be a bit more general, so it could be reused for e.g. generic type arguments or whatever.

mbeccati commented:
@nikic sure. But maybe we should move the discussion to internals and/or resurrect the RFC?

nikic commented:
@mbeccati Yeah, feel free to bring it up on internals again. If there is no disagreement over the APIs I hope we can land this without dragging through the whole RFC cycle.

mbeccati commented:
@nikic to be honest I tried a couple of times already without gathering any attention. Maybe if it comes from you we have better chances to land it ;)

morrisonlevi commented:
If you look at the RFC history I've probably edited it more than Sara has :)

Personally, I don't think we should include `isNull`, `isScalar`, etc. Furthermore, there is a proposal out for  union types would would basically require a new API *again* should it pass.

I honestly think our best solution is to ship without reflection support. Having a bad Reflection API is worse than not having one at all in my opinion; I mean this constructively. Our current API is not very good and this PR basically continues down the same path and just adds more methods.

morrisonlevi commented:
Some concrete feedback on this PR:

  - I don't think `isInstance` is appropriately named. I'm not sure `isObject` or `isClass` are any better. It's somewhat shocking to me that there isn't a well-known term for something that is either a class or an interface.
  - Why is there `isScalar` instead of multiple options, such as `isInt`, `isFloat`, etc?
  - I'm not sure why you created `typehint_reference` here; can you explain that?

nikic commented:
@morrisonlevi I don't think shipping without reflection is an option. I've already hit an issue with a mocking framework generating invalid code because it does not respect scalar and return typehints (due to lack of reflection support for them). Reflection is not just cosmetics.

Furthermore, I do not see why this would be incompatible with union types. For them isInstance(), isArray(), etc return false and isUnion() returns true - and getUnionTypes() or whatever returns an array of ReflectionTypes representing the individual members in the union. Similar handling is possible for generics.

mbeccati commented:
@morrisonlevi:
* following your reasoning, then `instanceof` is not appropriately named too.
* (string)$reflectionType already returns "int", "string", etc... why would we need specific isWhatever methods?
* I've reused Sara's code and expanded from there, so I can't explain that myself.

nikic commented:
@mbeccati Regarding the second point, you are of course right. However this begs the question why there are `isArray()` and `isCallable()` methods as these can also be checked by comparing the type string. Maybe we should drop those two and stick with just `__toString()`, `isNullable()` and `isInstance()`?

nikic commented:
After further discussion, the `isInstance()` method could be problematic if we introduce an `object` type in the future. If `isInstance()` is used to distinguish between "special" types and class/interface types, you wouldn't want it to return `true` for `object` - but that would be weird given it's name.

As such the suggestion was to call it `isClassOrInterface()` instead - we don't need to save characters, better be explicit.

nikic commented:
So, to summarize the current room 11 consensus:

```php
class ReflectionType {
    function __toString(): string;
    function isClassOrInterface(): bool;
    function allowsNull(): bool;
}
```

Another topic that came up is what `__toString()` returns for hints against `self` and `parent`. Your current implementation probably returns them directly, while it would likely be better to return the resolved names. Otherwise people will likely just have to manually resolve them. Furthermore this would allow us to resolve these names at compile-time in the future.

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -2690,6 +2754,116 @@ ZEND_METHOD(reflection_parameter, isVariadic)
+		case IS_DOUBLE:   RETURN_STRINGL("float", sizeof("float") - 1);
+		default:
+			php_error_docref(NULL, E_ERROR, "Unknown type hint: %d", (int)param->arg_info->type_hint);
+			RETURN_EMPTY_STRING();
```
Should use `EMPTY_SWITCH_DEFAULT_CASE()` here instead of the `E_ERROR` to assert that the switch enumerates all options.

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -2690,6 +2754,116 @@ ZEND_METHOD(reflection_parameter, isVariadic)
+	switch (param->arg_info->type_hint) {
+		case IS_ARRAY:    RETURN_STRINGL("array", sizeof("array") - 1);
+		case IS_CALLABLE: RETURN_STRINGL("callable", sizeof("callable") - 1);
+		case IS_OBJECT:   RETURN_STR(zend_string_copy(param->arg_info->class_name));
```
This needs to specially handle typehints on internal functions, see http://lxr.php.net/xref/PHP_TRUNK/ext/reflection/php_reflection.c#650 for an example. Would also be good to have a test for getting type info on an internal function.

mbeccati replied:
@nikic Right, the current code doesn't handle type hints on internal function / methods at all. I will check.

mbeccati commented:
@nikic I've fixed internal class support and EMPTY_SWITCH_DEFAULT_CASE().
@morrisonlevi now there's a reason to have a `typehint_reference` (we need both arg_info and fptr)

Re: room11 consensus, is that a thing? Internals or it didn't happen ;)

Seriously though that seems quite a big deviation from the original RFC, but I'll be happy to make the required changes if there's consensus also outside room11. About `self` and `parent`, I'm not sure. To be honest I don't actually like __toString() being the main way to get the type name; I also feel that it'd be useful to have a way to get both self/parent and the acual resolved class name if required.

morrisonlevi commented:
To clarify the Room 11 consensus: that includes Nikita Popov, Anthony Ferrara, myself and possibly Bob Weinand (can't remember if he agreed it was good or not; sorry if I'm putting words in your mouth, Bob). Github was not allowing me to post comments at the time, anyway.

The main intention of what Nikita has shared here is that it provides the baseline functionality without getting in the way of future changes or repeating past mistakes.

mbeccati commented:
@nikic @morrisonlevi I've updated the PR. For now I've added a specific test for self/parent to ensure they are returned as-is. Maybe we could add getReflectionClass() that returnes the resolved class/interface.

morrisonlevi commented:
Just do `new ReflectionClass((string) $type)`; that's my $0.02.

mbeccati commented:
@morrisonlevi so you don't think there is a case for keeping self/parent. Well, in that case:

https://github.com/mbeccati/php-src/commit/4d6d4f3

morrisonlevi commented:
I don't think we currently preserve `parent` and `self` on parameter types, do we?

mbeccati commented:
@morrisonlevi TBH I wasn't even aware that self/parent were legal parameter type "hints". In any case I've added some tests to cover them as well (to both branches).

mbeccati commented:
@morrisonlevi any further feedback?

morrisonlevi commented:
I think as long as we don't disclose `self` or `parent` in `ReflectionType` we are fine, I think. It's a new API and nobody will have behavior relying on `self` or `parent` being there.

morrisonlevi commented:
Going back to `isClassOrInterface`: this is too narrow. I still think `isInstance` was incorrect as well since it's not actually information about an *instance* but a *type*. Do you have any better ideas for names? `isClassLike`? `isTypeThatCanHaveMethods`? (okay, I'm not really proposing that last one)

nikic commented:
For the record, the reason why `isClassOrInterface()` is too narrow is that it would be a lie once we add other class-like structure, like `enum`s.

Personally I think `isClassLike()` is okay-ish, if a bit weird.

mbeccati commented:
@morrisonlevi @nikic I've merged the self/parent bit, and subsequently optimized it a little bit to avoid zend_string_init/copy + release in case self or parent are a match.

I haven't changed the method name yet, as the current best candidate `isClassLike()` still sounds a bit like a joke to me.

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -2690,6 +2761,91 @@ ZEND_METHOD(reflection_parameter, isVariadic)
+		if (len == sizeof("self") - 1 && !strncasecmp(name, "self", sizeof("self") - 1)) {
+			class_name = fptr->common.scope->name;
+		} else if (len == sizeof("parent") - 1 && !strncasecmp(name, "parent", sizeof("parent") - 1)) {
+			class_name = fptr->common.scope->parent->name;
```
This needs a check for whether `->parent` is NULL.

mbeccati replied:
oh yes, I was unsure if it could happen and I was supposed to ask, but I forgot. I will add it soon.

mbeccati replied:
@nikic Also, should I use zend_binary_strcasecmp ?

Danack commented:
So, @morrisonlevi suggested I should chime in on this if I wanted to see this merged, which he might regret doing in about 20 seconds...

If you want to stick with a is*** function, it would be better to have the method be isPrimitive and return true for the built-in types of arrays, resources, and the various scalars.

Avoiding the horrible name of isClassOrInterface should always be possible even when other types get added in the future:

```
if ($returnType->isPrimitive()) {
	//yay primitive.
}
else if ($returnType->isEnum()) {
	//yay enum
}
else {
	//everything else is a class or interface
}
```

However I don't think that is a great plan. It only covers two types of variable and I don't think it's really expandable. Although there are no pending RFCs to add other types, possibilities have been discussed quite extensively.

Having a method just return the actual type of the variable returned would be both more useful, as well as being expandable without having an explosion of methods.

```
class ReflectionType {
    function __toString() : string;
    function isNullable() : bool;
    function getType() : int; // returns a predefined type constant as below
}


TYPE_BOOLEAN
TYPE_INTEGER
TYPE_FLOAT
TYPE_STRING
TYPE_CALLABLE
TYPE_ARRAY
TYPE_OBJECT
TYPE_RESOURCE
TYPE_NULL
TYPE_ENUM
TYPE_UNION
```

Doing it like that should fit in well for both the return type reflection and the parameter reflection. Currently the parameter reflection has separate methods for `isArray` and `isCallable`. Using the ReflectionType::getType removes the need for adding a new method for each new possible type e.g. `isInteger`, `isBool`, `isFloat`.

This would also be expandable in the future, for example for union types. 

```
class ReflectionType {
    function __toString() : string;
    function isNullable() : bool;
    function getType() : int; // returns a predefined type constant
    function getUnionTypes(): array; //returns array of ReflectionTypes for functions that have a union 
                                     // return type. Returns null if this ReflectionType is not 
                                     // reflecting a union type. 
}
```


The only reason not to have the function be named getType is that it 'clashes' a little with ReflectionType, but they're different things. A ReflectionType is not the same as the type of a variable. If you really wanted to disambiguate them, the method could be called getVariableType.

btw I think TYPE_OBJECT would be correct, rather than TYPE_CLASS_OR_INTERFACE. The function is returning an object that will be either be implementing an interface, or be child class of a certain class; but that is a implementation detail that the function isn't aware of. The only thing the function knows, and so the reflection of the function knows, is that the variable returned is going to be an object.

nikic commented:
@Danack `TYPE_OBJECT` has the same issue that the already suggested `isInstance()` or `isObject()` would have: If we introduce an `object` typehint in the future (which is not unlikely, the name was already reserved), this type will be either confusing or misleading. Given that you would want to use this method / type to distinguish between special typehints and class ones, you would *not* want it to return true for an `object` typehint.

Furthermore, I'm not sure I understood how having a `getType()` method will improve things. What we need is the ability to distinguish between different *categories* of typehints (primitives like `int`, the class/interface/enum typehints and potential future things like union types or generics). The individual types within a single category can already be distinguished based on the string representation. Whether you write `$type == "int"` or `$type->getType() == ReflectionType::TYPE_INT` seems to be about the same, with the latter being more verbose and having a larger interface.

nikic commented:
@mbeccati @morrisonlevi Regarding `self` and `parent` types, I've recently become aware of two problems:
* For trait methods the `self` and `parent` types refer to the scope of the class using the trait, rather than the trait itself. This means that reflection directly on a trait (is this possible?) will not be able to resolve them to a specific type.
* For closures the `self` and `parent` types are subject to rebinding. While we can resolve them to concrete types for a particular closure, it might be valuable information that the types of the parameters will change after a rebind, so we might not want to resolve them.

Given how we won't be able to resolve self and parent for all cases, maybe we should stick with not resolving them in any case?

mbeccati commented:
@nikic makes sense. I can easily revert self+parent resolution.

morrisonlevi commented:
I propose `isBuiltin`. It would cover types "built in" to the language:

 - scalars (`bool`, `int`, `float`, `string`)
 - `array`
 - `callable`
 - `null`
 - `resource`

(The latter two are not currently permitted for type declarations, but if they were allowed then I think they'd fit into this category)

How do you feel about this?

nikic commented:
And I concur with that proposal ;) For the record, the idea behind this is as follows:

By adding something like `isClassOrInterface()` we're pretty much hardcoding against the current situation. If we add enums, that method name is broken. `isClassLike()`, while better, still has fundamentally the same issue. If we introduce some kind of type alias syntax like `type iterable = array|Traversable` then `isClassLike()` would be wrong as well. In the end the only thing we can distinguish is whether something is a special builtin type or if it isn't. Everything else we do not know statically.

The choice of `isBuiltin()` over `isPrimitive()` is that our docs have a pretty clear definition of what a primitive type is and stuff like `callable` or `void` doesn't fit in that category.

mbeccati commented:
@nikic @morrisonlevi Updated and rebased, even though it looks like github master is out-of-sync atm.

In any case, please do not merge this one, I can manually merge https://github.com/mbeccati/php-src/tree/reflection-typehints-sth-v6 which has @sgolemon 's original implementation plus a squashed commit of everything I did on top of her work.

morrisonlevi commented:
You can squash these and force-push to this branch; github will handle it properly. I doubt that anyone who has this branched checked out will care (or should be able to figure out how to get the latest).

morrisonlevi commented:
I would also recommend renaming `ext/reflection/tests/parameters_typehint.phpt` to `ext/reflection/tests/parameters_types.phpt`.

mbeccati commented:
@morrisonlevi sure, I didn't do it right away to make your review easier. It's squashed now and the test is called ReflectionType_001.phpt, to match the existing ones.

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -2630,6 +2661,46 @@ ZEND_METHOD(reflection_parameter, getClass)
+
+	RETVAL_BOOL((param->fptr->type == ZEND_INTERNAL_FUNCTION ?
+		((zend_internal_arg_info*)param->arg_info)->type_hint :
+		param->arg_info->type_hint) != 0);
```
There is no need to differentiate between internal/userland here (and in getType), as the structures are compatible (by design). So it's okay to directly check `param->arg_info->type_hint`.

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -2867,6 +2938,67 @@ ZEND_METHOD(reflection_parameter, isVariadic)
+			if (param->fptr->type == ZEND_INTERNAL_FUNCTION) {
+				RETURN_STRING(((zend_internal_arg_info*)param->arg_info)->class_name);
+			}
+			RETURN_STR(zend_string_copy(param->arg_info->class_name));
```
Can use `RETURN_STR_COPY` here.

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -305,6 +313,8 @@ static void reflection_free_objects_storage(zend_object *object) /* {{{ */
 			reference = (parameter_reference*)intern->ptr;
 			_free_function(reference->fptr);
+			/* fallthrough */
+		case REF_TYPE_TYPE:
```
Not totally sure here, but shouldn't `_copy_function` be called when invoking the factory and the `_free_function` call be done here? Otherwise the trampoline function might get freed.

mbeccati replied:
@nikic I'm afraid I'll have to leave further tests about `_copy_function` / `_free_function` to you, or I will need some pointers.

I've also added tests for ReflectionFunction($closure) and ReflectionMethod($closure, '__invoke') and the latter doesn't seem to work with return types. I'll look into it as soon as I have time.

nikic replied:
The problem with the return type was not on your side. I've just landed a change that should fix this and a similar problem for variadic parameters.

The fest for copy/free I had in mind is something like this:

```php
$closure = function(Test $x) {};
$rm = new ReflectionMethod($closure, '__invoke');
$rp = $rm->getParameters()[0];
$rt = $rp->getType();
unset($rm, $rp);
var_dump((string) $rt);
```

And then see if this causes memory errors in valgrind.

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -3379,6 +3511,45 @@ ZEND_METHOD(reflection_function, getShortName)
+{
+	reflection_object *intern;
+	zend_function *fptr;
+	struct _arg_info *ai;
```
This variable looks unused.

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -2630,6 +2661,46 @@ ZEND_METHOD(reflection_parameter, getClass)
 /* }}} */
 
+/* {{{ proto public bool ReflectionParameter::hasType()
+   Rethern whether parameter has a type hint */
```
nit "Rethern", same for `hasReturnType()`.

(Also for Levi's peace of mind maybe drop the "hint" ^^)

nikic commented on ext/reflection/php_reflection.c:
```diff
@@ -2630,6 +2661,46 @@ ZEND_METHOD(reflection_parameter, getClass)
+}
+/* }}} */
+
+/* {{{ proto public string ReflectionParameter::getType()
```
Return type here should be `ReflectionType` rather than `string`. Same for `getReturnType()`.

mbeccati commented:
@nikic fixed leak and all tests are now passing. Tbh I haven't digged into the trampoline stuff too much, but

https://github.com/mbeccati/php-src/commit/0e7821c3679c9fb3db6c1da2ea80f3dd26207799#diff-39e3ad3b3135c65cc314ca09cc835530L2989

was required to properly pick the class names. However the same check is missing from most of the similar snippets in other parts of the file, so I'm wondering if I did something wrong.

nikic commented:
Hm, I hadn't thought of the problem with the class name for via-handler functions. The check you added will not generally work, because the function `__invoke` proxies to could also be an internal function (if created via `getClosure`). Not sure how to fix that just now.

nikic commented:
@dstogov What do you think about this problem? Basically the comment in http://lxr.php.net/xref/PHP_TRUNK/Zend/zend_closures.c#210 is not true as far as Reflection is concerned. Do we need an extra flag to distinguish whether it's user or internal arginfo or do you see some other way to solve this?

nikic commented:
As far as I'm concerned you can merge this PR -- the via-handler issue is a tangential problem we can resolve later.

Majkl578 commented:
I think there should be a method to get the name of the type, without explicit string cast, like `ReflectionType::getName()`.

Also `ReflectionType::getClass() : ReflectionClass` might be very handy. (Similar to `ReflectionParameter::getClass()`).

mbeccati commented:
@nikic will do in the next few days.

@Majkl578 TBH, I find the requirement to use a string cast a bit weird too. But that's how the RFC worked and I wanted to deviate from it as little as possible.

morrisonlevi commented:
Just a reminder: this needs to be a minimal set of functionality for 7.0
that can be expanded and built onto later. It is intentionally small so as
to not interfere with other features such as Enums or Union types which may
not have a simple string name or ReflectionClass.

On Sun, Jun 7, 2015, 3:58 PM Matteo Beccati <notifications@github.com>
wrote:

> @nikic <https://github.com/nikic> will do in the next few days.
>
> @Majkl578 <https://github.com/Majkl578> TBH, I find the requirement to
> use a string cast a bit weird too. But that's how the RFC worked and I
> wanted to deviate from it as little as possible.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/php/php-src/pull/1190#issuecomment-109802828>.
>

mbeccati commented:
Merged in 11ece75291e94dff763f505ba0d6555537d979b9
